asdf_astropy/tests/test_integration.py

```diff
@@ -64,13 +65,63 @@ def test_no_astropy_import():
     at import time.
     """
 
-    keys = [k for k in sys.modules if k.startswith("asdf_astropy") or any(k.startswith(m) for m in _ASTROPY_MODULES)]
-    for key in keys:
-        del sys.modules[key]
-
-    from asdf_astropy import integration
-
-    integration.get_resource_mappings()
-    integration.get_extensions()
-
-    assert not any(k for k in sys.modules if any(k.startswith(m) for m in _ASTROPY_MODULES))
+    # If any astropy or asdf_astropy modules are already imported
+    # remove them from sys.modules so we can later in this test check
+    # if those modules were imported. We have to store these as
+    # many submodules will reference or use types defined in other
+    # submodules
+    previous_modules = {}
+    for name in sys.modules.copy():
+        if name.startswith("asdf_astropy") or any(name.startswith(m) for m in _ASTROPY_MODULES):
+            previous_modules[name] = sys.modules[name]
+            del sys.modules[name]
+
+    # Register a module finder that just raises an exception if
+    # one of the tracked astropy modules is imported
+    class _Finder:
+        def find_spec(self, modulename, path=None, target=None):
+            if any(modulename.startswith(m) for m in _ASTROPY_MODULES):
+                msg = f"attempt to import astropy submodule({modulename}) during integration"
+                raise Exception(msg)  # noqa: TRY002
+
+    sys.meta_path.insert(0, _Finder())
+
+    try:
+        from asdf_astropy import integration
+
+        integration.get_resource_mappings()
+        integration.get_extensions()
+
+        assert not any(k for k in sys.modules if any(k.startswith(m) for m in _ASTROPY_MODULES))
+    finally:
+        # Restore the previously imported modules. This is necessary as other code
+        # may have already used the modules we removed from the cache (astropy.constants
+        # defines some quantities).
+        for name in previous_modules:
+            sys.modules[name] = previous_modules[name]
+        if isinstance(sys.meta_path[0], _Finder):
+            sys.meta_path.pop(0)
```

Comment:
Most of this should be placed into a fixture as it distracts from what the actual test is. Instead do something like:
```python
@pytest.fixture()
def clean_astropy_imports():
    """Temporally unload all astropy modules used by asdf-astropy"""

    # If any astropy or asdf_astropy modules are already imported
    # remove them from sys.modules so we can later in this test check
    # if those modules were imported. We have to store these as
    # many submodules will reference or use types defined in other
    # submodules
    previous_modules = {}
    for name in sys.modules.copy():
        if name.startswith("asdf_astropy") or any(name.startswith(m) for m in _ASTROPY_MODULES):
            previous_modules[name] = sys.modules[name]
            del sys.modules[name]

    # Register a module finder that just raises an exception if
    # one of the tracked astropy modules is imported
    class _Finder:
        def find_spec(self, modulename, path=None, target=None):
            if any(modulename.startswith(m) for m in _ASTROPY_MODULES):
                msg = f"attempt to import astropy submodule({modulename}) during integration"
                raise Exception(msg)  # noqa: TRY002

    sys.meta_path.insert(0, _Finder())

    # Setup complete return to the test
    yield

    # Restore the previously imported modules. This is necessary as other code
    # may have already used the modules we removed from the cache (astropy.constants
    # defines some quantities).
    for name in previous_modules:
        sys.modules[name] = previous_modules[name]
    if isinstance(sys.meta_path[0], _Finder):
        sys.meta_path.pop(0)


@pytest.mark.usefixtures("clean_astropy_imports")
def test_no_astropy_import():
    """
    Confirm that none of the ASDF plugins import astropy modules
    at import time.
    """
    from asdf_astropy import integration

    integration.get_resource_mappings()
    integration.get_extensions()

    assert not any(k for k in sys.modules if any(k.startswith(m) for m in _ASTROPY_MODULES))
```

Reply:
Good idea! Updated in: 4b193cf7eeb70ffa27f27dfd9cd62004555352a7 (ruff wanted a leading `_` to the fixture because of PT004).